torch/csrc/jit/script/module.cpp

```diff
@@ -137,6 +122,249 @@ void Module::to_impl(
+
+// remove the first module argument, replacing any access of its
+// parameters/attributes with extra_ivalue input Slots that hold what value to
+// pass into the graph
+std::pair<std::shared_ptr<Graph>, std::vector<Slot>> lower_graph(
+    const ModulePtr& self,
+    Graph& g_,
+    size_t self_offset = 0) {
+  std::shared_ptr<Graph> g = g_.copy();
+  std::vector<Slot> extra_ivalues;
+  std::unordered_map<Slot, size_t> slot_to_offset;
+  struct ToScan {
+    ModulePtr mod;
+    Node* n;
+    size_t offset;
+  };
+  std::vector<ToScan> to_scan;
+  std::vector<Node*> to_clean; // nodes that should be dead at the end
+  std::vector<Node*> forks_edited; // we need to potentially de-dup inputs
+                                   // to the fork node
+
+  auto getOrAddSlot = [&](const Slot& slot) -> Value* {
+    auto it = slot_to_offset.find(slot);
+    if (it != slot_to_offset.end()) {
+      size_t ivalues_start = g->inputs().size() - extra_ivalues.size();
+      return g->inputs().at(ivalues_start + it->second);
+    }
+    extra_ivalues.emplace_back(slot);
+    slot_to_offset[slot] = extra_ivalues.size() - 1;
+    return g->addInput()->setType(slot.type());
+  };
+
+  auto self_value = g->inputs().at(self_offset);
+
+  for (Use use : self_value->uses()) {
+    to_scan.emplace_back(ToScan{self, use.user, use.offset});
+  }
+  while (to_scan.size() > 0) {
+    auto e = to_scan.back();
+    to_scan.pop_back();
+
+    // when we lambda lift forks, first-class modules may be passed across
+    // forks. This code recursively lowers the module in the fork call.
+    if (e.n->kind() == prim::fork) {
+      auto subgraph = e.n->g(attr::Subgraph);
+      std::vector<Slot> new_slots;
+      std::tie(subgraph, new_slots) = lower_graph(e.mod, *subgraph, e.offset);
+      e.n->g_(attr::Subgraph, subgraph);
+      for (const Slot& slot : new_slots) {
+        e.n->addInput(getOrAddSlot(slot));
+      }
+      e.n->removeInput(e.offset);
+      forks_edited.emplace_back(e.n);
+      continue;
+    }
+    if (e.n->kind() != prim::GetAttr) {
+      throw ErrorReport(e.n->getSourceLocation())
+          << "temporary: the only valid use of a module is looking up an attribute";
+    }
+    Slot slot(e.mod, e.mod->type()->getAttributeSlot(e.n->s(attr::name)));
+    if (ClassTypePtr c = e.n->output()->type()->cast<ClassType>()) {
+      if (c->name() == "Module") {
+        auto obj = slot.value().toObject();
+        for (Use use : e.n->output()->uses()) {
+          to_scan.emplace_back(ToScan{obj, use.user, use.offset});
+        }
+        to_clean.emplace_back(e.n);
+        continue;
+      }
+    }
+    e.n->output()->replaceAllUsesWith(getOrAddSlot(slot));
+    e.n->destroy();
+  }
+
+  while (to_clean.size() > 0) {
+    Node* n = to_clean.back();
+    AT_ASSERT(!n->hasUses());
+    n->destroy();
+    to_clean.pop_back();
+  }
+  AT_ASSERT(!self_value->hasUses());
+  g->eraseInput(self_offset);
+
+  // for(Node* fork : forks_edited) {
```

Comment:
looks like we're not cleaning up the duplicate inputs?

Reply:
This codepath never ended up being necessary. Initially I thought it would be because of another bug that accidentally added slots to the graph multiple times. Ill delete it.